eip-0017.md

```diff
@@ -0,0 +1,70 @@
+
+For example, a malicious whale tried to take advantage of this simple structure by stealing UI fees from SigmaUSD web interface developers for some period of time. This happened because the proxy contracts were simply only trying to protect users from malicious activities, not the dApp infrastructure.
+
+The same whale tried to prevent user's requests (minting/redeeming) from being executed by the assemble service by retuning their funds as soon as funds were broadcasted in the network. This happened also because of the same reasons.
+
+Generally, proxy contracts should be designed to:
+- prevent dApp developers or any other attacker from taking advantage of user's funds in any manner
+- preserve the integrity of the dApp by preventing attacks like the ones explained in the above examples.
+
+To achieve all of the above, the below contract structure is proposed as an example:
+
+```scala
+{
+  // dApp-specific part ensuring that user will receive what he is paying for
+  val properFundUsage = {
+    val userOUt = OUTPUTS(1)
+    userOut.propositionBytes == fromBase64("$userAddress") && // user must be the recipient
+      myOut.tokens(0)._1 == fromBase64("$scTokenId") && // user must receive SigmaUSD
+      myOut.tokens(0)._2 >= $scAmountL && // the amount of SigmaUSD must be at least what user is paying for
+      HEIGHT < $timestampL && // this part is always true (timestamp is the unix-timestamp at the time of the request), it will cause compiled address to differ everytime
+      OUTPUTS.size == 2 // only refund box and transaction fee box is needed
+  }
+  
+  // ensuring dApp integrity is preserved - any dApp specific condition to ensure designed procedures won't be violated
+  val UIFeeOk = OUTPUTS(2).propositionBytes == fromBase64("$implementor") && OUTPUTS.size == 4 // UI fee must go to UI devs not any random person who assembles the transaction
+  val properBank = OUTPUTS(0).tokens(2)._1 == fromBase64("$bankNFT") // the real bank box of the sigmaUSD protocol must be used so not any random person can behave as the bank box
+  val dAppWorksFine = properFundUsage && UIFeeOk && properBank
+
+  // in any case, whether assembler refuses to execute the request or the request fails for any reason, user must be able to get back his funds
+  val returnFunds = { 
+    val total = INPUTS.fold(0L, {(x:Long, b:Box) => x + b.value}) - $returnFee // only refund transactions's fee must be deducted from user's funds
+    OUTPUTS(0).value >= total && OUTPUTS(0).propositionBytes == fromBase64("$userAddress") // user must receive the appropriate amount
+    && (PK("$assemblerNodeAddr") || HEIGHT > $refundHeightThreshold) // either dApp-specific node can return user's funds or some time (block) has to be passed first. This is useful for many reasons.
+  }
+
+  sigmaProp(dAppWorksFine || returnFunds) // either dApp must work as it is supposed to or user's funds must be returned
```

Comment:
You are calling _**dAppWorksFine**_ in the OR manner.
dAppWorksFine is composed of _**properFundUsage && UIFeeOk && properBank**_
_**properFundUsage**_ contains (require thanks to the AND used) _**OUTPUTS.size == 2**_
_**UIFeeOk**_ contains (require thanks to the AND used) _**OUTPUTS.size == 4**_
Hence that means that only one or none member of tupple (_**properFundUsage**_, _**UIFeeOk**_) could be satisfied in some specific moment - because the _**OUTPUTS.size**_ has constant value during the contract execution.
This implies that the _**dAppWorksFine**_ could never be satisfied (AND used).

Reply:
yes this looks like a bug. I believe the `OUTPUTS.size == 2` should be part of `returnFunds` instead of `properFundUsage`

Reply:
Right, last-minute changes broke the contract, fixed.